### DISCUSSION TITLE
[Question] Many-to-Many Relationship

### DISCUSSION THREAD

**Discussion by j0nm1** [NilCoalescing/djangochannelsrestframework#107] [FAQ] [answered]
First of all thanks for this great library. It made a lot of stuff I have to handle much easier. But now I'm confronted with a hard-to-solve problem. I have a many-to-many relationship and want to subscribe to all the children. Just like the documentation (https://github.com/hishnash/djangochannelsrestframework#subscribing-to-a-filtered-list-of-models) shows. But I can't think of a way to implement the `groups_for_signal` without database calls.

Let's say my models look like this:

```
from django.db import models

class Publication(models.Model):
    title = models.CharField(max_length=30)

    class Meta:
        ordering = ['title']

    def __str__(self):
        return self.title

class Article(models.Model):
    headline = models.CharField(max_length=100)
    publications = models.ManyToManyField(Publication)

    class Meta:
        ordering = ['headline']

    def __str__(self):
        return self.headline
```
(https://docs.djangoproject.com/en/3.2/topics/db/examples/many_to_many/)

To get the `groups_for_signal` I would need to do something like this:

```
for publication in instance.publications.all():
    yield f"-publication__{publication.pk}"
```
But as stated in the documentation this doesn't seem to be a good idea. Is there a better way that I forgot?

**Comment by hishnash** [comment]
yer this is not a trival situation.

maybe what you could do is:

When subsceribing subscribe to all Articles (assume this is a small enough list... not in the 1000s)

1) Then for changes just emmit an event based on the article ID.

2) create a ManyToMany table and subscribe to changes on this (based on the publication id) 

In the consume when you detect a change to a ManyToMany table (eg a new article is created) subscribe to that article. Or if the many to many table is updated and the article is removed from the publication un-subscribe from the articles changes.

does this make any sense? happy to mock up how this migh look in code if you need?



(if you expect there to be many articles then we need to come up with a differnt solution)

Go through your code and ensure the many to many table is also included as a prefetch related (this way you could emmmit an event for every single publication without doing extra db quries).

**Comment by j0nm1** [comment]
Hi, thanks for the ideas.
Sadly I cannot predict how many Articles will be in the system, so there I think subscribing to all of them is not a really good idea either. But the point you made with the `prefetch related` seems really promising in my opinion. When I prefetch all `publications` then should be fine using the code snippet I provided without any additional db query (please correct me if I'm wrong).

**Comment by hishnash** [comment, marked as answer]
@j0nm1 I know it has been a long time since you question but I wanted to follow up for those reading this in the future.  

As of v1.3 we now have M2m support with model observers.   To enable this you do:

```python
@model_observer(Article, many_to_many=True)
```

this will result in you getting update notifications when the many to many relationship is updated.
